docker.md

```diff
@@ -0,0 +1,102 @@
+# Docker
+
+- [Docker](#docker)
+  - [Build and host in Google Cloud Artifact Registry](#build-and-host-in-google-cloud-artifact-registry)
+    - [Step 1: Create AR Docker repository](#step-1-create-ar-docker-repository)
+    - [Step 2: Build and upload Docker image](#step-2-build-and-upload-docker-image)
+      - [Option A: Build using only Dockerfile](#option-a-build-using-only-dockerfile)
+      - [Option B: Build using Dockerfile and build config](#option-b-build-using-dockerfile-and-build-config)
+    - [Step 3: Confirm Docker image built and uploaded](#step-3-confirm-docker-image-built-and-uploaded)
+
+Docker is a tool designed to make it easier to create, deploy, and run
+applications. If you don't know what it is or how to use it, the
+[official docs](https://docs.docker.com/get-started/) are a good starting point.
+
+## Build and host in Google Cloud Artifact Registry
+
+You can host your Docker images in a Google Cloud
+[Artifact Registry](https://cloud.google.com/artifact-registry/docs/docker/quickstart)
+(AR) repository. Here are the basic steps for building a Docker image using
+[Cloud Build](https://cloud.google.com/build/docs/quickstart-build) and
+subsequently pushing it to AR - the following assume that you want to host it in
+the `australia-southeast1` region, and that you already have a `Dockerfile` in
+the working directory.
+
+### Step 1: Create AR Docker repository
+
+```shell
+LOCATION="australia-southeast1"
+PROJECT="<project>"
+REPO="<repo-name>"
+
+gcloud artifacts repositories create ${REPO} \
+    --repository-format "docker" \
+    --location ${LOCATION} \
+    --description "My awesome Docker repo" \
+    --project ${PROJECT}
+```
+
+Verify that it was created:
+
+```shell
+gcloud artifacts repositories list \
+    --project ${PROJECT} \
+    --location ${LOCATION}
+```
+
+### Step 2: Build and upload Docker image
+
+There are two ways to build Docker images with Cloud Build and upload to AR. For
+most cases, a `Dockerfile` is the only file required. For more complex build
+workflows, a build config is recommended. In both cases, the image is
+automatically uploaded to AR upon successful build.
+
+#### Option A: Build using only Dockerfile
```

Comment:
This makes it sound like `gcloud builds submit` is our preferred way to build Docker images for AR.

While I find it handy to use for testing, it's actually not that convenient to integrate with GitHub Actions workflows.

In general, I'd like our repositories to build images automatically whenever something gets merged into the `main` branch. That's somewhat easier to do with standard Docker and you're not paying for a GitHub worker CPU to wait on Cloud Build, which itself also uses a CPU.

You only have to make sure to authenticate with `gcloud`, but that's fairly simple too:
https://github.com/populationgenomics/analysis-runner/blob/main/.github/workflows/hail_update.yaml#L33

What do you think?

Reply:
I really don't have a strong opinion, just figuring out things as I go (maybe we can even forget about this PR - just thought I'd start a docker + cloud build document for others to learn as well).

I spent a bit of time fiddling with the docker images for the SV workflows. I take it we want to host all those images on our AR instead of pulling them from the US region, right? There are two types of images there: those that don't get updated that often (e.g. for delly/manta/melt) and those that get updated frequently based on changes to the gatk-sv source. The infrequent ones don't require auto builds since they're based on external tools. I think the gatk-sv ones though would need to be updated fairly regularly, though not sure yet. There's also this script they have that I want to read through tomorrow that does all the builds automatically:
https://github.com/broadinstitute/gatk-sv/blob/master/scripts/docker/build_docker.py

Reply:
I think it's a great idea to add Docker-related documentation, I'm just not sold on Cloud Build yet :).

That script looks really interesting. I guess in our fork we'd adjust it to use AR instead of GCR and fix the Broad references? Then we could just invoke it from GitHub Actions, right?